src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs

```diff
@@ -470,6 +481,8 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
             return new DiagnosticWithProgrammaticSuppression(this, programmaticSuppressionInfo);
         }
 
+        internal virtual SuppressionInfo? SuppressionInfo { get { return null; } }
```

Comment:
It feels weird to now have an internal property `SuppressionInfo` and a public API `GetSuppressionInfo(Compilation)` on `Diagnostic`, both of them giving the same information: https://github.com/dotnet/roslyn/blob/693e68dbbb374db922eda88edd58d86693136876/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L343-L347

I think we should instead consider making this new property public and marking the `GetSuppressionInfo(Compilation)` API obsolete. I'll let @jaredpar and @dotnet/roslyn-compiler let you know their preference.

Reply:
@333fred can u comment here?

Reply:
I don't think I understand the motivation for this PR. Adding `SuppressionInfo` as a general property on `Diagnostic` doesn't seem like a particularly good idea to me, as whether a specific diagnostic is suppressed is dependent on the compilation. The PR description isn't particularly helping me either: can I get a summary of what the change is actually addressing?

Reply:
Hi @333fred the changes I made in this PR is to retrieve the suppression info from the diagnostic object then put it into the resulting SARIF 2 file.
Why `IsSuppressed` can be a general property and `SuppressionInfo` not? These values were retrieved at the same time but the latter was not used. I just attach it back to the `Diagnostic` object in `SuppressMessageAttributeState.ApplySourceSuppressions(Diagnostic diagnostic)`.
BTW, it's my first time to contribute to this project, please clarify it clearly.

Reply:
Looking at the implementation of `GetSuppressionInfo`, it seems like it was very intentionally not a property. Realizing attributes and members is very expensive in terms of memory, and we should try to avoid it as much as possible. Is there any reason this PR cannot use the existing method on `DiagnosticInfo`?

Reply:
It will take a bit of plumbing to bring the compilation to here in order to allow the suppression info to be realized for this case, but it's all on internal types so it should be doable. I do not want to see us eagerly realizing all the elements of `SuppressionInfo` if we do not need to.

Reply:
This should be removed.